### DISCUSSION TITLE
让 v2ray 运行在浏览器中是否可行？

### DISCUSSION THREAD

**Discussion by cnryb** [v2fly/v2ray-core#1329] [Ideas]
WebAssembly 是 [W3C](https://www.w3.org/community/webassembly/) 标准的一部分，这也就意味着大部分浏览器都会支持它。
感觉只要把 v2ray 编译出 WebAssembly 的版本，就能够让 v2ray 运行在浏览器中。

只是不知道把 v2ray-core 编译出 WebAssembly 的版本是否可行？

如果可以的话，那么我们就不必在设备上专门启动 v2ray 程序，直接打开一个指定的页面就可以享用互联网了吧。

---
以下是看起来可行的一些证据

在 [MDN](https://developer.mozilla.org/en-US/docs/WebAssembly/Concepts) 中我们可以看到， WebAssembly 设计的目的之一就是为其它语言提供一个高效的编译目标。
> It is not primarily intended to be written by hand, rather it is designed to be an effective compilation target for source languages like C, C++, Rust, etc.

golang 看起来也是支持 WebAssembly 的
https://github.com/golang/go/wiki/WebAssembly

WebAssembly官网中也把 VPN 列入了使用案例 
https://webassembly.org/docs/use-cases/

**Comment by frank-pian** [comment]
我也有这种想法，目前最直观的场景是在特斯拉车机上。由于特斯拉还没开放app开发，因此车机如果想翻墙必须依靠已经翻墙的热点，这样体验不佳，最好的办法还是浏览器内置功能。
但是目前问题是wasm是跟着网页走的，你没办法在别人的网页中嵌入wasm程序，如果创建一个单独的wasm app网站去代理访问其他网页，又会有跨域问题

**Comment by netheril96** [comment]
能跑可是没权限代理本机流量吧

**Comment by flyysquirrel** [comment]
https://github.com/shadowsocks/shadowsocks-rust/issues/778

**Comment by flyysquirrel** [comment]
调研了一下能否在 chrome extension 的 background 里运行 v2ray, 结论是不行:

在文档里虽然有这种接口
https://developer.chrome.com/docs/extensions/reference/sockets_tcpServer/ 
但其实只能用在 chrome app 里, (已经 deprecated 了)

**Reply by mydogshitgold** [reply]
While it is actually feasible to use v2ray-core as an extension, the extent of its functionality is confined solely to the browsers. There are a number of extensions in the web store that are known to proxy traffic within the browser.

**Reply by flyysquirrel** [reply]
thanks for letting me know the news
Could you please tell me the link of the extension? I can't find it

**Reply by mydogshitgold** [reply]
> Could you please tell me the link of the extension? I can't find it

Sorry, I didn't mean that there is an existing browser extension that uses v2ray.

What I meant was that web extensions can be used to proxy browser traffic. There are some known VPN providers that publish this kind of extension. You can try to reverse engineer them if you are curious about how they work.
